### PR TITLE
Allow multiple updates to the same field

### LINKED ISSUE #8226: Enable several update operations of the same field
In 8fc00a9c4bdd95bfa920c4757e711a3368eeb195 several operations of the same field (in the same update) were prohibited:
```lua
tarantool> s:update(1, {{'&', 2, 0x0F}, {'|', 2, 0xC0}})
---
- error: 'Field 2 UPDATE error: double update of the same field'
...
```
It seem to be not the most common case but sometimes it's natural to make several operations, like in the example above: drop a group of bits while setting some others.

Of course it must be done not ruining the original optimization that disabled this feature.

### PR DESCRIPTION
**Pull request by nshy**
Note that there is a distinct PR https://github.com/tarantool/tarantool/pull/8619 for "salad: make rope_erase erase more then a single element" patch. It will reviewed there.

Closes #8226

Lint failure is FP for typedef:
```
ERROR: Function name and return value type should be placed on different lines
#210: FILE: src/lib/salad/rope.h:661:
+typedef void (*rope_visit_f)(const rope_data_t data, size_t size, void *cb_arg);

total: 1 errors, 293 lines checked
```

### CODE REVIEW

File: src/lib/salad/rope.h
Comment (Gerold103) [inline src/lib/salad/rope.h:480]: Comments like these are not helpful and tend to outdate very quickly. Lets just drop them.

File: src/lib/salad/rope.h
Comment (Gerold103) [inline src/lib/salad/rope.h:467]: You seem to misuse `then` where `than` is needed. Please, check all the usages and fix, in the commit message too.

File: src/box/xrow_update.c
Comment (Gerold103) [inline src/box/xrow_update.c:169]: In the second commit you updated test-run submodule. I suppose it was an accident?

File: src/box/xrow_update_field.c
Comment (Gerold103) [inline src/box/xrow_update_field.c]: Doing that even when the operation is just one looks like a big overkill. Same for traversing the rope with a callback ("virtual" function call on the number of nodes). I suppose it would show some noticeable perf penalty for splice operations. But I will leave that to the core team to decide if this is fine.

File: src/box/xrow_update_field.c
Comment (Gerold103) [inline src/box/xrow_update_field.c:640]: If you want to describe the args, then use the doxygen format, please. Same in all the other new code.

File: src/box/xrow_update_field.h
Comment (Gerold103) [inline src/box/xrow_update_field.h:188]: This empty line now seems excess. The is just one member after it, not a group of 2.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch! I will return with more comments later.

File: src/box/xrow_update_array.c
Comment (Gerold103) [inline src/box/xrow_update_array.c]: Is that correct that this commit (`update: keep result of scalar operation in tree itself`) makes the code parse the scalar operation argument multiple times (AFAIK, it was parsed only one time before).

File: src/box/xrow_update_field.c
Comment (Gerold103) [inline src/box/xrow_update_field.c:327]: Lets use `/**` opener for out-of-function comments. Same in the other places.

File: src/box/xrow_update_field.c
Comment (Gerold103) [inline src/box/xrow_update_field.c:349]: I suggest to use `unreachable()`. Same in the other places.

File: src/box/xrow_update_field.h
Comment (Gerold103) [inline src/box/xrow_update_field.h:111]: We don't put comments on the same line as code (you might see violations in the old code, but it is legacy), lets move them above each of the constants. Or just drop them, they are obvious anyway.

File: src/box/xrow_update_field.h
Comment (Gerold103) [inline src/box/xrow_update_field.h:458]: `calar` -> `scalar`.

**Review by Gerold103 — COMMENTED** [review]
Next comments. I will continue the review a bit later.

File: test/box/update.result
Comment (Gerold103) [inline test/box/update.result]: Lets uncomment this and update the results.

File: test/box/update.result
Comment (Gerold103) [inline test/box/update.result]: Ditto.

File: test/engine/update.result
Comment (Gerold103) [inline test/engine/update.result:1361]: `array` -> `map`.

File: src/box/xrow_update_field.h
Comment (Gerold103) [inline src/box/xrow_update_field.h:267]: `weather` -> `wether`.

File: src/box/xrow_update_field.h
Comment (Gerold103) [inline src/box/xrow_update_field.h:270]: Lets make the argument point at `const`.

File: src/box/xrow_update_field.h
Comment (Gerold103) [inline src/box/xrow_update_field.h:282]: These are not needed - they can be included into the `default`.

File: src/box/xrow_update_route.c
Comment (Gerold103) [inline src/box/xrow_update_route.c:270]: Typo? Non scalar leaf bar is when length is > 0, not 0.

File: test/box/update.result
Comment (Gerold103) [inline test/box/update.result:1698]: Ok, I see now why you commented out the tests in the prev patch. It is not a problem to change the same test multiple times during the patchset if it is justified. Lets not comment anything out.

File: changelogs/unreleased/gh-8226-multiple-same-field-update.md
Comment (Gerold103) [inline changelogs/unreleased/gh-8226-multiple-same-field-update.md:3]: Lets at least mention that you mean tuple fields, not just some other fields. Also you need to write a docbot request and go through the commit messages to fix typos, add ticket references.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Good stuff, I don't have any big comments.

File: src/lib/salad/rope.h
Comment (nshy) [reply src/lib/salad/rope.h:480]: You mean the implementation part? I would keep the first paragraph and the return block - it is good to have comment what function do in general.

Implementation part is also useful IMHO. We can move it inside function.

File: src/box/xrow_update_array.c
Comment (nshy) [reply src/box/xrow_update_array.c]: Sorry, I don't see how it is more then a single parse. Please, explain)

File: src/box/xrow_update_array.c
Comment (nshy) [reply src/box/xrow_update_array.c:455]: So then we call `xrow_update_op_do_##op_type`. Every such function checks it's type and print error. We only follow this branch `item->field.type != XUPDATE_SCALAR && item->field.type != XUPDATE_NOP` when we try to apply scalar operation to non-scalar field so error message will be appropriate.

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:282]: I'd like instead to put `unreachable()` to default so we have all operation explicitly in the switch.

File: src/box/xrow_update_route.c
Comment (nshy) [reply src/box/xrow_update_route.c:270]: Looks correct.

Here `child.bar.path_len` is referenced as length. This length can be 0 for scalar and non-scalar leafs. The branch above handles scalar so this comment is for non-scalar leaf. (I rephrased the previous comment as scalar leaf is now handled explicitly.) 

May be I'd better put () right after leaf word.

File: test/box/update.result
Comment (nshy) [reply test/box/update.result:1698]: AFAIR at least some of the tests leads to assertion. I'll then comment only such tests.

File: src/lib/salad/rope.h
Comment (nshy) [reply src/lib/salad/rope.h:467]: Fixed.

File: src/box/xrow_update.c
Comment (nshy) [reply src/box/xrow_update.c:169]: Fixed.

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:188]: Done.

File: test/box/update.test.lua
Comment (nshy) [reply test/box/update.test.lua:631]: The thing is if I keep this test unchanged then after the patch I'll get:
```
---
- error: 'Tuple field 3 (g) type does not match one required by operation: expected
    array, got map'
...
```
It is not from update itself. Update will produce `{a=34}` in the second field of the tuple. But tuple format states second field should be an array. So test become a bit strange.

File: src/box/xrow_update_field.c
Comment (nshy) [reply src/box/xrow_update_field.c]: Discussed with @alyapunov Decided we should try to avoid performance degradation. We probably can do the simplest thing - use old structure until there is a second update - in that case convert it to rope.

File: test/box/update.test.lua
Comment (Gerold103) [reply test/box/update.test.lua:631]: It might be that this test in its new form won't cover the crash fix done in the same commit where the test was introduced. That is why I would propose to keep it as is even if it fails weirdly. It just shouldn't crash. It is not expected to succeed AFAIR.

File: src/box/xrow_update_array.c
Comment (Gerold103) [reply src/box/xrow_update_array.c]: You added here `xrow_update_mp_read_scalar()` which parses `item->field.data`. This code can be done by multiple update operations having the same path, right? Which means that every operation reaching this place will read the scalar field again. Wouldn't it?

File: src/box/xrow_update_array.c
Comment (Gerold103) [reply src/box/xrow_update_array.c:455]: Ok, I see now, thanks. It is worth adding a short comment about that here.

File: changelogs/unreleased/gh-8226-multiple-same-field-update.md
Comment (p7nov) [inline changelogs/unreleased/gh-8226-multiple-same-field-update.md:3]:
```suggestion
* Allowed multiple updates to the same tuple field in a single call (gh-8226).
```

**Review by p7nov — APPROVED** [review]
Please see my suggestion on the changelog.

File: src/lib/salad/rope.h
Comment (nshy) [reply src/lib/salad/rope.h:480]: I guess I'll keep then. The thing is the API do not check operation validity (assertions at the beginning of `rope_erase`) so the only possible error is OOM. Looks like it good to know it. In particular it is used in the next patch in assertions for `xrow_update_rope_*` return value.

Regularly I don't specify error reasons it the comment too.

File: src/box/xrow_update_array.c
Comment (nshy) [reply src/box/xrow_update_array.c]: It is not quite so. After this patch the second update will still hit `item->field.type != XUPDATE_NOP` raising `xrow_update_err_double` error. And after the next patch which touches the same place the parse will be done only on the first time when `item->field.type == XUPDATE_NOP`. After that type is changed to `XUPDATE_SCALAR` and we only do the update operation itself without parsing.

File: src/box/xrow_update_field.c
Comment (nshy) [reply src/box/xrow_update_field.c]: After optimization degradation is gone:
A = 1.57
D = 0.029

File: test/box/update.test.lua
Comment (nshy) [reply test/box/update.test.lua:631]: Fixed.

File: src/box/xrow_update_field.c
Comment (nshy) [reply src/box/xrow_update_field.c:327]: Fixed.

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:111]: Fixed to separate line for the comment.

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:458]: Fixed.

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:829]: Dropped.

File: src/box/xrow_update_array.c
Comment (nshy) [reply src/box/xrow_update_array.c:450]: Done.

File: src/box/xrow_update_array.c
Comment (nshy) [reply src/box/xrow_update_array.c:455]: Done.
`/* Just stub for non scalar field. op_do_ * will fail on it. */`

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:267]: Fixed.

File: src/box/xrow_update_field.h
Comment (nshy) [reply src/box/xrow_update_field.h:270]: Done.

File: src/box/xrow_update_route.c
Comment (nshy) [reply src/box/xrow_update_route.c:270]: Changed to `(path length is 0)`.

File: changelogs/unreleased/gh-8226-multiple-same-field-update.md
Comment (nshy) [reply changelogs/unreleased/gh-8226-multiple-same-field-update.md:3]: Done.

File: test/engine/update.result
Comment (nshy) [reply test/engine/update.result:1361]: Fixed.

File: test/box/update.result
Comment (nshy) [reply test/box/update.result:1698]: Adjusted patch set a bit to avoid temporary assertions and as a result avoid commenting tests out.

File: test/box/update.result
Comment (nshy) [reply test/box/update.result]: Done. Add a comment for temporary result:
`Currenly result is not correct. It is temporary and will be fixed later.`

File: test/box/update.result
Comment (nshy) [reply test/box/update.result]: Moved the below hunk to a later patch so now I don't need commenting these tests.
```diff
-               if (old_token.type == JSON_TOKEN_END) {
-                       /*
-                        * The bar path ended. It means the new operation either
-                        * is trying to update exactly the same field, or go
-                        * even deeper.
-                        */
-                       xrow_update_err_double(new_op);
-                       return NULL;
-               }
```
In the previous version of the patch set I deleted hunk too early -  it works until the patch where it deleted now.

File: test/engine/update.result
Comment (nshy) [reply test/engine/update.result:1219]: Moved tests to `box/update.test.lua`.

Not sure I need to add tests testing first field or field in the middle. New related code seems to be orthogonal to position in the array.

May be we can change tests to update field in the middle to check that last field in not corrupted/overwritten somehow? Not sure I need to test first/middle/last element for every case

File: src/lib/salad/rope.h
Comment (Gerold103) [reply src/lib/salad/rope.h:480]: You still need to convert it to doxygen format.

**Review by Gerold103 — COMMENTED** [review]
In the first commit message you still have the `*TODO* Add ticket reference.` stub.

File: test/engine/update.result
Comment (Gerold103) [reply test/engine/update.result]: > Not sure I need to add tests testing first field or field in the middle. New related code seems to be orthogonal to position in the array.

I wouldn't say so. There is at least one commit where you touch the tail update data in a struct, which in your tests is almost always empty, it seems.

> May be we can change tests to update field in the middle to check that last field in not corrupted/overwritten somehow?

You could reuse the big complex tuples in `box/update.test.lua` (lines 297, 514) created specifically for that. The tests would remain the same, but would operate on a more complex tuple.

**Comment by nshy** [comment]
I still plan to merge the first commit in a different PR. It's review is not even started though).

File: src/box/xrow_update_field.c
Comment (nshy) [reply src/box/xrow_update_field.c]: Fixed.

File: test/engine/update.result
Comment (nshy) [reply test/engine/update.result]: > I wouldn't say so. There is at least one commit where you touch the tail update data in a struct, which in your tests is almost always empty, it seems.

If you are referencing commit `update: make '=' NOP in update tree instead of scalar` - this should be covered by existing tests.

> You could reuse the big complex tuples in box/update.test.lua (lines 297, 514) created specifically for that. The tests would remain the same, but would operate on a more complex tuple.

I would say large tuple will make it harder to follow testing case. In the above places it justified as JSON path is tested.

If I use a complex tuple should I in every test case pick up a different place? This seems a bit a different test flavor, more like a random test. If it is a desired then I'm OK to change but it looks a bit of work)

I guess it is not justified currently that there is a field before updated and there is no field after it. It is a remnant of the first version where I used `space:update()`. May be I should just add an after field?

Comment (Gerold103) [reply test/engine/update.result]: Adding an after field would be good. And also keep the existing tests updating the last field as well. For such complicated code there can't be too many tests.

> every test case pick up a different place?

If you want. I would just recommend testing non-last and last field updates.

> but it looks a bit of work)

This isn't much more complicated than your current tests really. You just have to check results one time in the .result file.

Anyway, I will leave that to the next reviewer.

**Review by Gerold103 — APPROVED** [review]
Thanks for the fixes!

**Comment by nshy** [comment]
Thanx for the review!

File: test/engine/update.result
Comment (nshy) [reply test/engine/update.result]: Added last field for tuples in test. Typical tuple now looks like this `t = box.tuple.new({7, {a = 1}, 11}) `.

**Pull request merged**
